<a href="https://colab.research.google.com/github/kelngu/python-for-math/blob/main/Notebook%20%231/P4M_Notebook1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#  Computing with Python:  an Introduction

Welcome to programming!

What is the difference between *Python* and a calculator?  We begin this first lesson by showing how Python can be used **as** a calculator, and we move into one of the most important programming structures -- the **loop**.  Loops allow computers to carry out repetetive computations, with just a few commands.

To navigate this notebook, you might wish to click on the Table of Contents icon (near top-left of the Google Colab window).  Then you can expand the sections to work on each one.

Remember to save your work frequently on your personal GitHub repository.

# Python as a calculator

Different kinds of data are stored as different *types* in Python.  For example, if you wish to work with integers, your data is typically stored as an *int*.  A real number might be stored as a *float*.  There are types for booleans (True/False data), strings (like "Hello World!"), and many more we will see.  

A more complete reference for Python's numerical types and arithmetic operations can be found in the [official Python documentation](https://docs.python.org/3/library/stdtypes.html).  The [official Python tutorial](https://docs.python.org/3/tutorial/introduction.html) is also a great place to start.

Python allows you to perform arithmetic operations:  addition, subtraction, multiplication, and division, on numerical types.  The operation symbols are `+`, `-`, `*`, and `/`.   Evaluate each of the following cells to see how Python performs operations on *integers*.  

To evaluate the cell, click anywhere within the cell to select it and use the keyboard shortcut *Shift-Enter* to evaluate.  But as you go through this and later lessons, try to *predict* what will happen when you evaluate the cell before you hit Shift-Enter.  Don't just click click click!  Think for a few seconds or a minute, and then Shift-Enter.

In [ ]:
2 + 3

5

In [ ]:
2 * 3

6

In [ ]:
5 - 11

-6

In [ ]:
5.0 - 11

-6.0

In [ ]:
5 / 11

0.45454545454545453

In [ ]:
6 / 3

2.0

In [ ]:
5 // 11

0

In [ ]:
6 // 3

2

The results are probably not too surprising, though the last two require a bit of explanation.  Python *interprets* the input number 5 as an *int* (integer) and 5.0 as a *float*.  "Float" stands for "floating point number," which are decimal approximations to real numbers.  The word "float" refers to the fact that the decimal (or binary, for computers) point can float around (as in 1.2345 or 12.345 or 123.45 or 1234.5 or 0.00012345).  There are deep computational issues related to how computers handle decimal approximations, and you can [read about the IEEE standards](https://en.wikipedia.org/wiki/IEEE_754) if you're interested.

Python enables different kinds of division.  The single-slash division in Python 3.x gives a floating point approximation of the quotient.  That's why `5 / 11` and `6 / 3` both output floats.  On the other hand, `5 // 11` and `6 // 3` yield integer outputs (rounding down) -- this is useful, but one has to be careful!

In fact the designers of Python changed their mind.  **This tutorial assumes that you are using Python 3.x.**  If you are using Python 2.x, the command `5 / 11` would output zero.

In [ ]:
-12 // 5  # What will this output?  Guess before evaluating!

-3

Why use integer division `//` and why use floating point division?  In practice, integer division is typically a faster operation.  So if you only need the rounded result (and that will often be the case), use integer division.  It will run much faster than carrying out floating point division then manually rounding down.

Observe that floating point operations involve approximation.  The result of `5.0/11.0` might not be what you expect in the last digit.  Over time, especially with repeated operations, *floating point approximation* errors can add up!

You might be wondering about the little [XX] and [XX] prompts in the code cells above.  What is their purpose?  Guess what the following line will do.

In [ ]:
Out[4] + Out[5]

-12.0

Cool, huh?  It's nice to have a record of previous computations, especially if you don't want to type something again.

Python allows you to group expressions with parentheses, and follows the order of operations that you learn in school.

In [ ]:
(3 + 4) * 5

35

In [ ]:
3 + (4 * 5)

23

In [ ]:
3 + 4 * 5   #  What do you think will be the result?  Remember PEMDAS?

23

Now is a good time to try a few computations of your own, in the empty cell below.  You can type any Python commands you want in the empty cell.  If you want to insert a new cell into this notebook, just hover your cursor in the space between cells in Colab.

In [ ]:
2.0/7 #  An empty cell.  Have fun!


0.2857142857142857

For number theory, *division with remainder* is an operation of central importance.  Integer division provides the quotient, and the operation `%` provides the remainder.  It's a bit strange that the percent symbol is used for the remainder, but this [dates at least to the early 1970s](https://softwareengineering.stackexchange.com/questions/294297/in-what-programming-language-did-the-use-of-the-percent-sign-to-mean-modulo) and has become standard across computer languages.

In [ ]:
23 // 5  # Integer division

4

In [ ]:
23 % 5  # The remainder after division

3

Note in the code above, there are little "comments".  To place a short comment on a line of code, just put a hashtag `#` at the end of the line of code, followed by your comment.

Python gives a single command for division with remainder.  Its output is a *tuple*.

In [ ]:
divmod(23,5)

(4, 3)

In [ ]:
type(divmod(23,5))

tuple

All data in Python has a type, but a common complaint about Python is that types are a bit concealed "under the hood".  But they are not far under the hood!  Anyone can find out the type of some data with a single command.

In [ ]:
type(3)

In [ ]:
type(3.0)

In [ ]:
type('Hello')

In [ ]:
type([1,2,3])

The key to careful computation in Python is always being *aware of the type* of your data, and *knowing* how Python operates differently on data of different types.

In [ ]:
3 + 3

6

In [ ]:
3.0 + 3.0

6.0

In [ ]:
'Hello' + 'World!'

'HelloWorld!'

In [ ]:
[1,2,3] + [4,5,6]

[1, 2, 3, 4, 5, 6]

In [ ]:
3 + 3.0

6.0

In [ ]:
3 + 'Hello!'  # Uh oh!

TypeError: ignored

In [ ]:
#  An empty cell.  Have fun!
#  Try operating on ints, floats, and strings, with different operations.  Which ones work?  How?

As you can see, addition (the `+` operator) is interpreted differently in the contexts of numbers, strings, and lists.  The designers of Python allowed us to add *numbers* of different types:  if you try to operate on an *int* and a *float*, the *int* will typically be *coerced* into a float in order to perform the operation.  But the designers of Python did not give meaning to the addition of a number with a string, for example.  That's why you probably received a *TypeError* after trying to add a number to a string. 

On the other hand, Python does interpret *multiplication* of a natural number with a string or a list.

In [ ]:
3 * 'Hello!'

'Hello!Hello!Hello!'

In [ ]:
0 * 'Hello!'

''

In [ ]:
2 * [1,2,3]

[1, 2, 3, 1, 2, 3]

Can you create a string with 100 A's (like `AAA...`)?  Use an appropriate operation in the cell below.

In [ ]:
100 * ('A') #  Practice cell


'AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA'

Exponents in Python are given by the `**` operator.  The following lines compute 2 to the 1000th power, in two different ways.

In [ ]:
2**1000

10715086071862673209484250490600018105614048117055336074437503883703510511249361224931983788156958581275946729175531468251871452856923140435984577574698574803934567774824230985421074605062371141877954182153046474983581941267398767559165543946077062914571196477686542167660429831652624386837205668069376

In [ ]:
2.0**1000

1.0715086071862673e+301

As before, Python interprets an operation (`**`) differently in different contexts.  When given integer input, Python evaluates `2**1000` **exactly**.  The result is a large integer.  A nice fact about Python, for mathematicians, is that it handles exact integers of arbitrary length!  Many other programming languages (like C++) will give an error message if integers get too large in the midst of a computation.  

New in version 3.x, Python implements long integers without giving signals to the programmer or changing types.  In Python 2.x, there were two types: *int* for somewhat small integers (e.g., up to $2^{31}$) and *long* type for all larger integers.  Python 2.x would signal which type of integer was being used, by placing the letter "L" at the end of a long integer.  Now, in Python 3.x, the programmer doesn't really see the difference.  There is only the *int* type.  But Python still optimizes computations, using hardware functionality for arithmetic of small integers and custom routines for large integers.  The programmer doesn't have to worry about it most of the time.

For scientific applications, one often wants to keep track of only a certain number of significant digits (sig figs).  If one computes the floating point exponent `2.0**1000`, the result is a decimal approximation.  It is still a float.  The expression "e+301" stands for "multiplied by 10 to the 301st power", i.e., Python uses *scientific notation* for large floats.

In [ ]:
type(2**1000)

int

In [ ]:
type(2.0**1000)

float

In [ ]:
#  An empty cell.  Have fun!


Now is a good time for reflection.  Double-click in the cell below to answer the given questions.  Cells like this one are used for text rather than Python code.  Text is entered using *markdown*, but you can typically just enter text as you would in any text editor without problems.  Press *shift-Enter* after editing a text cell to complete the editing process.  

### Exercises

1.  What data types have you seen, and what kinds of data are they used for?  Can you remember them without looking back?

2.  How is division `/` interpreted differently for different types of data?

3.  How is multiplication `*` interpreted differently for different types of data?

4.  What is the difference between 100 and 100.0, for Python?

Double-click this text cell to edit it, and answer the exercises.  This may be graded, so please complete all questions!  Write in clear, complete, and concise sentences.

### Solutions


1. We have seen three different data types: data that is stored as an integer (int), data that is stored as a floating point number (float), and data that is stored as a boolean.

2. There are a couple of scenarios where division with one backslash `/` can be interpreted differently for different types of data. \
If we divide an integer with another integer, for example, `5 / 2`, the output will result in a floating point number, which is `2.5`. \
When we divide a floating number with another floating number, on the other hand, for example, `6.0 / 4.0`, the output will also result in a floating point number, which is `1.5`, \
Even when we divide a floating number with an integer, for example, `6.0 / 4`, the output will also result in a floating point number, which is `1.5`. \
Thus, when using one backslash `/ ` for division, the output seems to always result in a floating point number, even if the numbers being divide can be either both integers, both floating numbers, or a mixture of both integers and floating numbers. 

3. There are a couple of scenarios where multiplication with one asterisk ` * ` can be interpreted differently for different types of data. \
If we multiply an integer with another integer, for example, `5 * 2`, the output will result in an integer, which is `10`. \
When we multiply a floating number with another floating number, for example, `6.0 * 4.0`, the output will result in a floating point number instead,  which is `24.0`. \
When we multiply a floating number with an integer, on the other hand, for example, `6.0 * 4`, the output will also result in a floating point number, which is `24.0`, even though an integer is present in our input. \
Thus, when using one asterisk ` * ` for multiplication, the output seems to depend on the input. If the input includes at least one floating point number, then the multiplication output will be a floating point number, even if an integer is present in the input. Otherwise, if both of the numbers are integers, then the output will then be integers. 

4. For Python, the difference between `100` and `100.0` is how the two numbers, which are equal to each other, are stored. While `100` is stored as an integer, `100.0` is stored as a floating point number instead. The difference in how Python stores their data lies within certain operations. As seen in the third problem in this exercise, if the we multiply two integers, for instance `100 * 100`, we will achieve another integer `10000`. But if we decide to multiply one integer and a floating point number `100 * 100.0` or multiply two floating point numbers `100.0 * 100.0`, we will yield a floating point number `10000.0`. The way Python stores their data can dictate how we calculate our numbers and how we write our script, which will vary by what we need.  

# Calculating with booleans

A *boolean* (type *bool*) is the smallest possible piece of data.  While an *int* can be any integer, positive or negative, a *boolean* can only be one of two things:  *True* or *False*.  In this way, booleans are useful for storing the answers to yes/no questions.  

Questions about (in)equality of numbers are answered in Python by *operations* with numerical input and boolean output.  Here are some examples.  A more complete reference is [in the official Python documentation](https://docs.python.org/3/library/stdtypes.html#boolean-operations-and-or-not).

In [ ]:
3 > 2

True

In [ ]:
type(3 > 2)

bool

In [ ]:
10 < 3

False

In [ ]:
2.4 < 2.4000001

True

In [ ]:
32 >= 32

True

In [ ]:
32 >= 31

True

In [ ]:
2 + 2 == 4

True

Which number is bigger:  $23^{32}$ or $32^{23}$?  Use the cell below to answer the question!

In [ ]:
23**32 < 32**23 #  Write your code here.
                # If boolean output is "True," then 23**32 < 32**23. Else, 23**32 > 32**23.

False

The expressions `<`, `>`, `<=`, `>=` are interpreted here as **operations** with numerical input and boolean output.  The symbol `==` (two equal symbols!) gives a True result if the numbers are equal, and False if the numbers are not equal.  An extremely common typo is to confuse `=` with `==`.  But the single equality symbol `=` has an entirely different meaning, as we shall see.

Using the remainder operator `%` and equality, we obtain a divisibility test.

In [ ]:
63 % 7 == 0  # Is 63 divisible by 7?

True

In [ ]:
101 % 2 == 0  # Is 101 even?

False

Use the cell below to determine whether 1234567890 is divisible by 3.

In [ ]:
1234567890 % 3 == 0 # Your code goes here.
                    # If boolean output is "True," then 1234567890 is divisible by 3. Else, 1234567890 is NOT divisible by 3.


True

Booleans can be operated on by the standard logical operations: and, or, not.  In ordinary English usage, "and" and "or" are conjunctions, while here in *Boolean algebra*, "and" and "or" are operations with Boolean inputs and Boolean output.  The precise meanings of "and" and "or" are given by the following **truth tables**.

    
| and | True | False |
|-----|------|-------|
| **True** | True | False |
| **False** | False | False|

<br>
  
| or | True | False |
|-----|------|-------|
| **True** | True | True |
| **False** | True | False|

In [ ]:
True and False

False

In [ ]:
True or False

True

In [ ]:
True or True

True

In [ ]:
not True

False

Use the truth tables to predict the result (True or False) of each of the following, before evaluating the code.

In [ ]:
(2 > 3) and (3 > 2)

False

In [ ]:
(1 + 1 == 2) or (1 + 1 == 3)

True

In [ ]:
not (-1 + 1 >= 0)

False

In [ ]:
2 + 2 == 4

True

In [ ]:
2 + 2 != 4  # For "not equal", Python uses the operation `!=`.

False

In [ ]:
2 + 2 != 5  # Is 2+2 *not* equal to 5?

True

In [ ]:
not (2 + 2 == 5)  # The same as above, but a bit longer to write.

True

Experiment below to see how Python handles a double or triple negative, i.e., something with a `not` `not`.

In [ ]:
not (not (2 + 2 == 5)) # Experiment here.


False

In [ ]:
not (not (not (2 + 2 == 5))) # Experiment here.

True

Python does give an interpretation to arithmetic operations with booleans and numbers.  Try to guess this interpretation with the following examples.  Change the examples to experiment!

In [ ]:
False * 100 # False = 0.

0

In [ ]:
True + 13 # True = 1.

14

This ability of Python to interpret operations based on context is a mixed blessing.  On one hand, it leads to handy shortcuts -- quick ways of writing complicated programs.  On the other hand, it can lead to code that is harder to read, especially for a Python novice.  Good programmers aim for code that is easy to read, not just short!

The [Zen of Python](https://www.python.org/dev/peps/pep-0020/) is a series of 20 aphorisms for Python programmers.  The first seven are below.

> Beautiful is better than ugly.

> Explicit is better than implicit.

> Simple is better than complex.

> Complex is better than complicated.

> Flat is better than nested.

> Sparse is better than dense.

> Readability counts.

### Exercises

1.  Did you look at the truth tables closely?  Can you remember, from memory, what `True or False` equals, or what `True and False` equals?  

2.  How might you easily remember the truth tables?  How do they resemble the standard English usage of the words "and" and "or"?

3.  If you wanted to know whether a number, like 2349872348723, is a multiple of 7 but **not** a multiple of 11, how might you write this in one line of Python code?

4.  You can chain together `and` commands, e.g., with an expression like `True and True and True` (which would evaluate to `True`).  You can also group booleans, e.g., with `True and (True or False)`.  Experiment to figure out the order of operations (`and`, `or`, `not`) for booleans.

5.  The operation `xor` means "exclusive or".  Its truth table is: `True xor True = False` and `False xor False = False` and `True xor False = True` and `False xor True = True`.  How might you implement `xor` in terms of the usual `and`, `or`, and `not`?



###  Solutions

(Edit here to give solutions to the exercises)

1. `True or False = True`. \
`True and False = False`. 

2. In ordinary English usage, `and` and “or” are conjunctions, while here in Boolean Algebra, “and” and “or” are operations with Boolean inputs and outputs. \
From what we learned in Math 100, we could easily remember the truth table based on logic and patterns. Suppose we have two propositions, “p” and “q” such that we form an “or” statement “p or q.” For the “or” statement to be false, both “p” and “q” must be false; otherwise, if either “p,” “q,” or both propositions are true, then the “or” statement is true. \
Suppose we still have two propositions, “p” and “q” such that we form an “and” statement “p or q.” For the “and” statement to be true, both “p” and “q” must be true; otherwise, if either “p,” “q,” or both propositions are false, then the “and” statement is false.

3. `(x % 7 == 0) and (not (x % 11 == 0))`

4. Based on the experiments I conducted, I figured out that the order of operations for booleans in Python is: \
(1) not \
(2) and \
(3) or.

5. Let “p” and “q” be two propositions such that we form an “xor” statement “p xor q.” To implement the “xor” statement in terms of the usual “and,” “or,” and “not,” we can say that: \
`(p or q) and (not (p and q))`. \
The code below shows the "xor" statement in terms of “and,” “or,” and “not” with in four different cases:

In [ ]:
(True or True) and (not (True and True)) # p=true, q=true; True xor True = False = (true or true) and (not (true and true))

False

In [ ]:
(True or False) and (not (True and False)) # p=true, q=false; True xor False = True = (true or false) and (not (true and false))

True

In [ ]:
(False or True) and (not (False and True)) # p=false, q=true; False xor True = True = (false or true) and (not (false and true))

True

In [ ]:
(False or False) and (not (False and False)) # p=false, q=false; False xor False = False = (false or false) and (not (false and false))

False

# Declaring variables

A central feature of programming is the declaration of variables.  When you declare a variable, you are *storing* data in the computer's *memory* and you are assigning a *name* to that data.  Both storage and name-assignment are carried out with the *single* equality symbol =.

In [ ]:
e = 2.71828

With this command, the float 2.71828 is stored somewhere inside your computer, and Python can access this stored number by the name "e" thereafter.  So if you want to compute "e squared", a single command will do.

In [ ]:
e * e

7.3890461584

In [ ]:
type(e)

float

You can use just about any name you want for a variable, but your name *must* start with a letter, *must* not contain spaces, and your name *must* not be an existing Python word.  Characters in a variable name can include letters (uppercase and lowercase) and numbers and underscores `_`.  

So `e` is a valid name for a variable, but `type` is a bad name.  It is very tempting for beginners to use very short abbreviation-style names for variables (like `dx` or `vbn`).  But resist that temptation and use more descriptive names for variables, like `difference_x` or `very_big_number`.  This will make your code readable by you and others!

There are different style conventions for variable names.  We use lowercase names, with underscores separating words,  roughly following [Google's style conventions](https://google.github.io/styleguide/pyguide.html#Python_Style_Rules) for Python code.

In [ ]:
my_number = 17

In [ ]:
my_number < 23

True

After you declare a variable, its value remains the same until it is changed.  You can change the value of a variable with a simple assignment.  After the above lines, the value of my_number is 17.

In [ ]:
my_number = 3.14

This command reassigns the value of my_number to 3.14.  Note that it changes the type too!  It effectively overrides the previous value and replaces it with the new value.

Often it is useful to change the value of a variable *incrementally* or *recursively*.  Python, like many programming languages, allows one to assign variables in a self-referential way.  What do you think the value of S will be after the following four lines?

In [ ]:
S = 0
S = S + 1
S = S + 2
S = S + 3
print(S)

6


The first line `S = 0` is the initial declaration:  the value 0 is stored in memory, and the name S is assigned to this value.

The next line `S = S + 1` looks like nonsense, as an algebraic sentence.  But reading = as **assignment** rather than **equality**, you should read the line `S = S + 1` as assigning the *value* `S + 1` to the *name* `S`.  When Python interprets `S = S + 1`, it carries out the following steps.

1.  Compute the value of the right side, `S+1`.  (The value is 1, since `S` was assigned the value 0 in the previous line.)
2.  Assign this value to the left side, `S`.  (Now `S` has the value 1.)

Well, this is a slight lie.  Python probably does something more efficient, when given the command `S = S + 1`, since such operations are hard-wired in the computer and the Python interpreter is smart enough to take the most efficient route.  But at this level, it is most useful to think of a self-referential assignment of the form `X = expression(X)` as a two step process as above.

1.  Compute the value of `expression(X)`.
2.  Assign this value to `X`.

Now consider the following three commands.

In [ ]:
my_number = 17
new_number = my_number + 1
my_number = 3.14

What are the values of the variables my_number and new_number, after the execution of these three lines?

To access these values, you can use the *print* function.

In [ ]:
print(my_number)
print(new_number)

3.14
18


Python is an *interpreted* language, which carries out commands line-by-line from top to bottom.  So consider the three lines

``` python
my_number = 17
new_number = my_number + 1
my_number = 3.14
```

Line 1 sets the value of my_number to 17.  Line 2 sets the value of new_number to 18.  Line 3 sets the value of my_number to 3.14.  But Line 3 does *not* change the value of new_number at all.

(This will become confusing and complicated later, as we study mutable and immutable types.)

### Exercises

1.  What is the difference between `=` and `==` in the Python language?

2.  If the variable `x` has value `3`, and you then evaluate the Python command `x = x * x`, what will be the value of `x` after evaluation?

3.  Imagine you have two variables `a` and `b`, and you want to switch their values.  How could you do this in Python?

4.  Kepler's third law states that for a planet in circular orbit around an object of mass $M$, one has $4 \pi^2 r^3 = G M t^2$.  We can use this to estimate the mass of the sun, from other astronomically observable quantities.  Look up $G$ (the gravitational constant, estimated by experiment on Earth) and $r$ (the distance from the Earth to the sun, in meters).  Compute $t$, the number of seconds it takes for the Earth to go around the sun (365.25 days).  Finally use all of this to estimate $M$, the mass of the sun.  Your solution should use 5-10 lines of Python code.

### Solutions

(Use this space to work on the exercises.  Place more code and text cells above as needed.)

**1.** `=` is used as an assignment of values to said variable.  \
`==` is used as a comparison between variables, values, or strings. 

**2.** After evaluation, `x` will have a value of `9`, as seen in the following code below:

In [ ]:
x = 3           # declares x to be 3.
x = x * 3       # the new value of x will be the product of the old value of x, which is 3, and 3.
print(x)        # prints value of x after evaluation.

9


**3.** The following code below switches the values of a and b:

In [ ]:
a = 3               # declare value for a, can be any number but not the same as b to see switching
b = 4               # declare value for b, can be any number but not the same as a to see switching

temp = a            # set temporary variable to equal value of a
a = b               # a equals to b, indicates variable a switching to variable b
b = temp            # b now equals to temporary variable which equals to value of a, indicates variable b switching to variable b

print(a)            # prints value of a after switching.
print(b)            # prints value of b after switching.

4
3



**4.** Before estimating the mass of the sun, we needed to find the values of `G`, `r`, and `t`. The values of `G` and `r` were found on Google, while we computed `t` as the number of seconds in a year, which is product of the amount of seconds in a minute, the amount of minutes in an hour, the amount of hours in a day, and the amount of days in a year. Now to estimate the mass of the sun, we needed to isolate the variable of `M`, which gives us our estimated value of the mass of the sun below. Note that we needed to import a math library in order to include a more accurate value of `pi` into our mass equation. 

In [ ]:
import math                                                             # imports math library for pi constant   

r = 1.496 * (10**(11))                                                  # declares value of r as distance from Earth to Sun
G = 6.67408 * (10**(-11))                                               # declares value of G as gravitational constant
t = 60 * 60 * 24 * 365.25                                               # secs * mins * hrs * days = num of seconds in a year

M = (4 * (math.pi**2) * (r**3)) // (G * (t**2))                         # calculates the mass of the sun

print('The estimated mass of the sun is: ' + str(M))                    # prints the mass of the sun after evaluation

The estimated mass of the sun is: 1.9886354409672198e+30


# Lists and ranges

Python stands out for the central role played by *lists*.  A *list* is what it sounds like -- a list of data.  Data within a list can be of any type.  Multiple types are possible within the same list!  The basic syntax for a list is to use brackets to enclose the list items and commas to separate the list items.

In [ ]:
type([1,2,3])

list

In [ ]:
type(['Hello',17])

list

There is another type called a *tuple* that we will use less often.  Tuples use parentheses for enclosure instead of brackets.

In [ ]:
type((1,2,3))

tuple

There's another list-like type in Python 3, called the `range` type.  Ranges are kind of like lists, but instead of plunking every item into a slot of memory, ranges just have to remember three integers:  their *start*, their *stop*, and their *step*.    

The `range` command creates a range with a given start, stop, and step.  If you only input one number, the range will ***start at zero*** and use ***steps of one*** and will stop ***just before*** the given stop-number.

One can create a list from a range (plunking every term in the range into a slot of memory), by using the `list` command.  Here are a few examples.

In [ ]:
type(range(10)) # Ranges are their own type, in Python 3.x.  Not in Python 2.x!

range

In [ ]:
list(range(10)) # Let's see what's in the range.  Note it starts at zero!  Where does it stop?

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

A more complicated two-input form of the range command produces a range of integers **starting at** a given number, and **terminating before** another given number.

In [ ]:
list(range(3,10))

[3, 4, 5, 6, 7, 8, 9]

In [ ]:
list(range(-4,5))

[-4, -3, -2, -1, 0, 1, 2, 3, 4]

This is a common source of difficulty for Python beginners.  While the first parameter (-4) is the starting point of the list, the list ends just before the second parameter (5).  This takes some getting used to, but experienced Python programmers grow to like this convention.

The *length* of a list can be accessed by the len command.

In [ ]:
len([2,4,6])

3

In [ ]:
len(range(10))  # The len command can deal with lists and ranges.  No need to convert.

10

In [ ]:
len(range(10,100)) # Can you figure out the length, before evaluating?

90

The final variant of the range command (for now) is the *three-parameter* command of the form `range(a,b,s)`.  This produces a list like `range(a,b)`, but with a "step size" of `s`.  In other words, it produces a list of integers, beginning at `a`, increasing by `s` from one entry to the next, and going up to (but not including) `b`.  It is best to experiment a bit to get the feel for it!

In [ ]:
list(range(1,10,2))

[1, 3, 5, 7, 9]

In [ ]:
list(range(11,30,2))

[11, 13, 15, 17, 19, 21, 23, 25, 27, 29]

In [ ]:
list(range(-4,5,3))

[-4, -1, 2]

In [ ]:
list(range(10,100,17))

[10, 27, 44, 61, 78, 95]

This can be used for descending ranges too, and observe that the final number b in range(a,b,s) is not included.

In [ ]:
list(range(10,0,-1))

[10, 9, 8, 7, 6, 5, 4, 3, 2, 1]

How many multiples of 7 are between 10 and 100?  We can find out pretty quickly with the range command and the len command (to count).

In [ ]:
list(range(10,100,7))  # What list will this create?  It won't answer the question...

[10, 17, 24, 31, 38, 45, 52, 59, 66, 73, 80, 87, 94]

In [ ]:
list(range(14,100,7))  # Starting at 14 gives the multiples of 7.

[14, 21, 28, 35, 42, 49, 56, 63, 70, 77, 84, 91, 98]

In [ ]:
len(range(14,100,7))  # Gives the length of the list, and answers the question!

13

### Exercises

1.  If `a` and `b` are integers, what is the length of `range(a,b)`?  Express your answer as a formula involving `a` and `b`.

2.  Use a list and range command to produce the list `[1,2,3,4,5,6,7,8,9,10]`.

3.  Create the list `[1,2,3,4,5,1,2,3,4,5,1,2,3,4,5,1,2,3,4,5,1,2,3,4,5]` with a single list and range command and another operation.

4.  How many multiples of 3 are there between 300 and 3000 (including 300 and 3000)?

### Solutions

(Use this space to work on the exercises.  Place more code and text cells above as needed.)

**1.** If `a` and `b` are integers, we can measure the length of `range(a,b)` with the following formula: \
`range(a,b) = b - a`. 

**2.** We could use a list and range command to produce the list `[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]`.

In [ ]:
list(range(1,11,1))

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

**3.** To create the list `[1,2,3,4,5,1,2,3,4,5,1,2,3,4,5,1,2,3,4,5,1,2,3,4,5]` with a single list and range command and another operation, we know that `list(range(1,6,1))` will give us list `[1,2,3,4,5]`. From what we are given, we see that list `[1,2,3,4,5]` is repeated 5 times, so we could multiple the list by 5, which gives us: 

In [ ]:
5*list(range(1,6,1))

[1, 2, 3, 4, 5, 1, 2, 3, 4, 5, 1, 2, 3, 4, 5, 1, 2, 3, 4, 5, 1, 2, 3, 4, 5]

**4.** From the code below, we found that there are 901 multiples of 3 between 300 and 3000. 

In [ ]:
len(range(300,3001,3))      # Gives the length of the list to calculate number of multiples of 3, range from 300 to 3001 (3001 is excluded), increments by 3 for multiples of 3

901

# Iterating over a range

Computers are excellent at repetitive reliable tasks.  If we wish to perform a similar computation, many times over, a computer a great tool.  Here we look at a common and simple way to carry out a repetetive computation:  the "for loop".  The "for loop" *iterates* through items in a list or range, carrying out some action for each item.  Two examples will illustrate.

In [ ]:
for n in [1,2,3,4,5]:
    print(n*n)

1
4
9
16
25


In [ ]:
for s in ['I','Am','Python']:
    print(s + "!")

I!
Am!
Python!


The first loop, **unraveled**, carries out the following sequence of commands.

In [ ]:
n = 1
print(n*n)
n = 2
print(n*n)
n = 3
print(n*n)
n = 4
print(n*n)
n = 5
print(n*n)

1
4
9
16
25


But the "for loop" is more efficient *and* more readable to programmers.  Indeed, it saves the repetition of writing the same command `print(n*n)` over and over again.  It also makes transparent, from the beginning, the range of values that `n` is assigned to.  

When you read and write "for loops", you should consider how they look unravelled -- that is how Python will carry out the loop.  And when you find yourself faced with a repetetive task, you might consider whether it may be wrapped up in a for loop.  Most of the time, you should try to write code that *never repeats* the same task.  If there is repetition, wrap it in a loop.

Try to unravel the loop below, and predict the result, before evaluating the code.

In [ ]:
P = 1
for n in range(1,6):
    P = P * n
print(P)

120


This might have been difficult!  So what if you want to trace through the loop, as it goes?  Sometimes, especially when debugging, it's useful to inspect every step of the loop to see what Python is doing.  We can inspect the loop above, by inserting a print command within the *scope* of the loop.

In [ ]:
P = 1
for n in range(1,6):
    P = P * n
    print("n is",n,"and P is",P)
print(P)

n is 1 and P is 1
n is 2 and P is 2
n is 3 and P is 6
n is 4 and P is 24
n is 5 and P is 120
120


Here we have used the *print* command with strings and numbers together.  In Python 3.x, you can print multiple things on the same line by separating them by commas.  The "things" can be strings (enclosed by single or double-quotes) and numbers (int, float, etc.).

In [ ]:
print("My favorite number is",17)

My favorite number is 17


If we unravel the loop above, the linear sequence of commands interpreted by Python is the following.

In [ ]:
P = 1
n = 1
P = P * n
print("n is",n,"and P is",P)
n = 2
P = P * n
print("n is",n,"and P is",P)
n = 3
P = P * n
print("n is",n,"and P is",P)
n = 4
P = P * n
print("n is",n,"and P is",P)
n = 5
P = P * n
print("n is",n,"and P is",P)
print (P)

n is 1 and P is 1
n is 2 and P is 2
n is 3 and P is 6
n is 4 and P is 24
n is 5 and P is 120
120


Let's analyze the loop syntax in more detail.  
```python
P = 1
for n in range(1,6):
    P = P * n  # this command is in the scope of the loop.
    print("n is",n,"and P is",P)  # this command is in the scope of the loop too!
print(P)
```
The "for" command ends with a colon `:`, and the **next two** lines are indented.  The colon and indentation are indicators of **scope**.  The *scope* of the for loop begins after the colon, and includes all indented lines.  The *scope* of the for loop is what is repeated in every step of the loop (in addition to the reassignment of `n`).  

In [ ]:
P = 1
for n in range(1,6):
    P = P * n  # this command is in the scope of the loop.
    print("n is",n,"and P is",P)  # this command is in the scope of the loop too!
print(P)

n is 1 and P is 1
n is 2 and P is 2
n is 3 and P is 6
n is 4 and P is 24
n is 5 and P is 120
120


If we change the indentation, it changes the scope of the for loop.  Predict what the following loop will do, by unraveling, before evaluating it.

In [ ]:
P = 1
for n in range(1,6):
    P = P * n
print("n is",n,"and P is",P)
print(P)

n is 5 and P is 120
120


Scopes can be nested by nesting indentation.  What do you think the following loop will do?  Can you unravel it?

In [ ]:
for x in [1,2,3]:
    for y in ['a', 'b']:
        print(x,y)

1 a
1 b
2 a
2 b
3 a
3 b


How might you create a nested loop which prints `1 a` then `2 a` then `3 a` then `1 b` then `2 b` then `3 b`?  Try it below.

In [ ]:
for y in ['a','b']:                                         # Insert your loop here.
  for x in [1,2,3]:
    print(x,y)                    

1 a
2 a
3 a
1 b
2 b
3 b


Among popular programming languages, Python is particular about indentation.  Other languages indicate scope with open/close braces, for example, and indentation is just a matter of style.  By requiring indentation to indicate scope, Python effectively removes the need for open/close braces, and enforces a readable style.

We have now encountered data types, operations, variables, and loops.  Taken together, these are powerful tools for computation!  Now complete the following exercises for more practice.

## Exercises

1.  Describe how Python interprets division with remainder when the divisor and/or dividend is negative.
2.  What is the remainder when $2^{90}$ is divided by $91$?
3.  How many multiples of 13 are there between 1 and 1000?
4.  How many *odd* multiples of 13 are there between 1 and 1000?
5.  What is the sum of the numbers from 1 to 1000?
6.  What is the sum of the squares, from $1 \cdot 1$ to $1000 \cdot 1000$?
7.  Euler proved that
$$\frac{1}{1^4} + \frac{1}{2^4} + \frac{1}{3^4} + \cdots = \frac{\pi^4}{C},$$
for some positive integer $C$.  Use Python to guess what $C$ is.

## Solutions



1.   If both the dividend and the divisors are positive, then we treat the output to be the remainder. But if we were to really consider what is going on with the `%` operator, we notice that what it is really doing is finding the remainder by using the modulo operation. When the either dividend or the divisor is negative, then the output could potentially be negative, but actually be completely different than when the both the dividend and the divisors are positive. Thus, Python interprets division with remainder when the divisor and/or dividend is negative as though if `a` is an integer divident and `b` is some integer divisor such that if `q` is the remainder while `r` is the remainder, then `b*q + r = a` implying that `r = a - b*q` by definition of the modulus operation. We notice that if either `a` is negative, then `r = -(a + b*q)`, and if `b` is negative, then `r = (a + b*q)`, indicating why Python interprets a negative divisor or negative dividend differently than when both divisor and dividend are positive.   



In [46]:
-5 % 3

1

In [45]:
-5 % -3

-2

In [ ]:
5 % -3 

-1



2.   We could use the operator `%` to find the remainder of `2**90` is divided by `91`. Thus, the remainder when `2**90` is divided by `91` is `64`. 



In [ ]:
2**90 % 91 # 2^90 is divided by 90, the remainder is given as such

64

3. We could set the range from `13` to `1000` with increments of `13` because we are looking for multiples of `13`. By using the `len` command, between `1` and `1000`, there are `76` multiples of `13`. 

In [ ]:
len(list(range(13,1001,13)))      # multiples of 10 from the range of 13 to 1000, "len" counts multiples

76

4. We could set the range from `13` to `1000` with increments of `13` because we are looking for multiples of `13`. We then created a for loop, which checks to see if each multiple is even or odd; if the multiple is odd, then that number is added into the list `odd`, but if the multiple is even, then nothing happens to that number, and we move on from it. That `odd` list accumulates all of the odd multiples of 13 between `1` and `1000`. Thus, after implementing the `len` command on the list `odd`, there are `38` odd multiples of `13`.

In [ ]:
odd=[]                                    # defines odd to be list of all odd multiples of 13

for n in range(13,1001,13):               # iterates through all multiples of 13 in the range of 13 to 1000
  if n % 2 == 1:                          # if multiple has a remainder of 1, then multiple is odd
    odd.append(n)                         # if multiple is odd, then store multiple in list "odd"

print(odd)                                # print final list "odd" after all iterations are complete
len(odd)                                  # length of list "odd" which is all odd multiples of 13 in the range of 13 to 1000

[13, 39, 65, 91, 117, 143, 169, 195, 221, 247, 273, 299, 325, 351, 377, 403, 429, 455, 481, 507, 533, 559, 585, 611, 637, 663, 689, 715, 741, 767, 793, 819, 845, 871, 897, 923, 949, 975]


38

5. We set a base `p=0`, in which from there, we create a for loop that adds the number `p` with all of the numbers `n` in the range of `1` to `100`. Hence, the sum of the numbers from `1` to `1000` is `5050`. 

In [ ]:
p = 0                                     # begin with a base at P = 0

for n in range(1,101):                    # range from 1 to 100, 101 is not included in range 
  p += n                                  # means "sum of p" + n which is then stored as p

print(p)                                  # prints the sum of all of the numbers from 1 to 100

5050


6. We also set a base p=0, but instead, we create a for loop that adds the number p with all of the numbers `n**2` in the range of 1 to 100. Therefore, the sum of the squares, from  `1*1`  to  `1000*1000` is `338350`.

In [ ]:
p = 0                                     # begin with a base at P = 0

for n in range(1,101):                    # range from 1 to 100, 101 is not included in range 
  p += n**2                               # means "sum of p" + n^2 which is then stored as p

print(p)                                  # prints the sum of all of the numbers from 1 to 100

338350


7. For some positive integer  `C`, we used a for loop to first add all of the numbers on the left side of th given equation; we used a large range of `1` to `100000` so that we could get closer to that number `C`. Because it is some number equaling to `pi^4 / C`, then using Algebra, we would need to divide `pi^4` by `p` to get `C`. Thus, using Python, we guessed that `C`  is about `90`.

In [ ]:
import math                                   # imports math library for pi constant

p = 0                                         # begin with a base at P = 0

for n in range(1,100001):                     # range from 1 to 100001, 100001 is not included in range 
  p += (1/(n**4))                             # "sum of p" + (1/n^4) which is then stored as p 

C = ((math.pi)**4) / p                        # Isolate/Solve for C, sum of fracs = pi^4/C -> C = pi^4/sum of fracs

print(p)                                      # prints an estimate sum of all fractions of the left side 
print(C)                                      # prints an estimate of C

1.082323233710861
90.00000000002302


# Explorations

Now that you have learned the basics of computation in Python and loops, we can start exploring some interesting mathematics!  We are going to look at approximation here -- some ancient questions made easier with programming.

## Exploration 1:  Approximating square roots.



We have seen how Python can do basic arithmetic -- addition, subtraction, multiplication, and division.  But what about other functions, like the square root?  In fact, Python offers a few functions for the square root, but that's not the point.  How can we compute the square root using only basic arithmetic?  

Why might we care?

1.  We might want to know the square root of a number with more precision than the Python function offers.
2.  We might want to understand how the square root is computed... under the hood.
3.  Understanding approximations of square roots and other functions is important, because we might want to approximate other functions in the future (that aren't pre-programmed for us).

Here is a method for approximating the square root of a number $X$.

1.  Begin with a guess $g$.
2.  Observe that $g \cdot (X / g) = X$.  Therefore, among the two numbers $g$ and $(X/g)$, one will be less than or equal to the square root of X, and the other will be greater than or equal to the square root.
3.  Take the average of $g$ and $(X/g)$.  This will be closer to the square root than $g$ or $X/g$ (unless your guess is exactly right!) 
4.  Use this average as a new guess... and go back to the beginning.

Now implement this in Python to approximate the square root of 2.  Use a loop, so that you can go through the approximation process 10 times or 100 times or however many you wish.  Explore the effect of different starting guesses.  Would a change in the averaging function improve the approximation?  How quickly does this converge?  How does this change if you try square roots of different positive numbers?

Write your code (Python) and findings (in Markdown cells) in a readable form.  Answer the questions in complete sentences.



### Solutions

In [49]:
#  Start your explorations here!

x = 3
g = 30

for i in range(9):
  g = (g + (x/g)) / 2
print(g)

import math 
print(math.sqrt(x))

1.7320508075688772
1.7320508075688772


## Exploration 2:  Approximating e and pi.

### Approximating e using the Taylor Series Expansion



Now we approximate two of the most important constants in mathematics:  e and pi.  There are multiple approaches, but e is pretty easy with the series expansion of e^x.  First, approximate e by evaluating the Taylor series expansion of e^x at x=1.  I.e., remember from calculus that
$$e^x = 1 + x + \frac{1}{2} x^2 + \frac{1}{3!} x^3 + \frac{1}{4!} x^4 + \cdots.$$
How many terms are necessary before the float stabilizes?  Use a loop, with a running product for the factorials and running sums for the series.

#### Solutions

In [107]:
#  Approximate e here.
import math                             # import math library to for euler's constant

fact = 1                                # start factorial at 1                             
sumTaylor = 1                           # start Taylor series sum with 1

for n in range(1,18):                   # begins stabilizing at 17, no need to go beyond 18
  fact *= n                             # factorial = 1*(current n iteration), next iteration is next factorial = (factorial from previous iteration) * (current n)
  sumTaylor += 1/fact                   # Taylor = Taylor + (1/fact current iteration), next iteration is next Taylor = (Taylor from previous iteration) + (1/factorial current iteration)

print('The math library calculates the value of Eulers Constant to be: ' + str(math.e))                           
print('In order for the float to stabilize, the Taylor Series needs to expand to ' + str(len(list(range(1,18)))) + ' terms in order to approximate Eulers constant accurately.')
print('Thus, we approximated Eulers constant to be: ' + str(sumTaylor))                        

The math library calculates the value of Eulers Constant to be: 2.718281828459045
In order for the float to stabilize, the Taylor Series needs to expand to 17 terms in order to approximate Eulers constant accurately.
Thus, we approximated Eulers constant to be: 2.7182818284590455


Based on the Python script, the for loop had to go iterate through a range of `range(1,18)`, thus 17 terms are necessary in order for the float to stabilize. 

### Approximating pi using $arctan(1) = \pi/4$

Next we will approximate pi, which is much more interesting.  We can try a few approaches.  For a series-approach (like e), we need a series that converges to pi.  A simple example is the arctangent atan(x).  Recall (precalculus!) that $atan(1) = \pi/4$.  Moreover, the derivative of $atan(x)$ is $1 / (1+x^2)$.  

1.  Figure out the Taylor series of $1 / (1+x^2)$ near $x=0$.  Note that this is a geometric series! 

2.  Figure out the Taylor series of $atan(x)$ near $x=0$ by taking the antiderivative, term by term, of the above.

3.  Try to estimate $\pi$ with this series, using many terms of the series.

#### Solutions

In [103]:
#  Approximate pi here!
import math                                                                     # import math library to calculate pi

arctan1 = 0                                                                     # begins base arctan1 = 0
num = -1                                                                        # alternating series, start base numerator at -1 so that first term is -1*-1=1 which is positive

for den in range(1,10000002,2):                                                 # den range of odd numbers from 1 to 10000001 based on taylor series formula
  num *= -1                                                                     # multiple num by -1 to act like alternating series 
  arctan1 += (num**(den))/den                                                   # by taylor series formula, taylor = last taylor + (num^den)/(den), should alternate - and +

print('The math library calculates the value of pi to be: ' + str(math.pi))                                                                 
print('We found arctan(1) to be: ' + str(arctan1))
print('Using the fact that arctan(1) = pi/4, we estimated pi to be: ' + str(arctan1*4))   
print('From our Python script, it seems that Taylor Series Expansion uses ' + str(len(list(range(1,10000002,2)))) + 'terms in order for our approxation of pi to be as accurate as possible.')                                              

The math library calculates the value of pi to be: 3.141592653589793
We found arctan(1) to be: 0.7853982133974349
Using the fact that arctan(1) = pi/4, we estimated pi to be: 3.1415928535897395
From our Python script, it seems that Taylor Series Expansion uses 5000001terms in order for our approxation of pi to be as accurate as possible.


### Approximating pi using $\pi/4 = 4 \cdot arctan(1/5) - arctan(1 / 239)$.

Now we'll accelerate things a bit.  There's a famous formula of Machin (1706) who computed the first hundred digits of $\pi$.  We'll use his identity:

$\pi/4 = 4 \cdot atan(1/5) - atan(1 / 239)$.  

This isn't obvious, but there's a tedious proof using sum/difference identities in trig.

Try using this formula now to approximate $\pi$, using your Taylor series for $atan(x)$.  It should require fewer terms.

#### Solutions

In [102]:
#  Approximate pi more quickly here!
import math                                                                     # import math library to calculate pi

arctan5 = 0                                                                     # begins base arctan5 = 0
num5 = -(1/5)                                                                   # alternating series, start base numerator at -1/5 so that first term is -1/5*-1=1/5 which is positive

arctan239 = 0                                                                   # begins base arctan1 = 0
num239 = -(1/239)                                                               # alternating series, start base numerator at -1 so that first term is -1/239*-1=1/239 which is positive

for den in range(1,22,2):                                                       # den range of odd numbers from 1 to 21 based on taylor series formula
  num5 *= -1                                                                    # multiple num by -1 to act like alternating series 
  arctan5 += ((num5)**(den))/den                                                # by taylor series formula, taylor = last taylor + (num^den)/(den), should alternate - and +
  num239 *= -1                                                                  # multiple num by -1 to act like alternating series 
  arctan239 += ((num239)**(den))/den                                            # by taylor series formula, taylor = last taylor + (num^den)/(den), should alternate - and +

pi_approx = 4*arctan5 - arctan239                                               # pi/4 = 4*arctan5 - arctan239

print('The math library calculates the value of pi to be: ' + str(math.pi))  
print('We found arctan(5) to be: ' + str(arctan5))
print('We found arctan(239) to be: ' + str(arctan239))
print('Using the fact that arctan(5) - arctan(239) = pi/4, we estimated pi to be: ' + str(pi_approx*4))
print('From our Python script, it seems that the Taylor Series Expansion uses ' + str(len(list(range(1,22,2)))) + ' terms in order for our approxation of pi to be as accurate as possible, which is MUCH less than just using arctan(1) = pi/4.') 

The math library calculates the value of pi to be: 3.141592653589793
We found arctan(5) to be: 0.1973955598498808
We found arctan(239) to be: 0.0041840760020747225
Using the fact that arctan(5) - arctan(239) = pi/4, we estimated pi to be: 3.141592653589794
From our Python script, it seems that it takes 11 terms in order for our approxation of pi to be as accurate as possible, which is MUCH less than just using arctan(1) = pi/4.


### Archimedes' Approximation of pi

Now let's compare this to **Archimedes' method**.  Archimedes approximated pi by looking at the perimeters $p(n)$ and $P(n)$ of ($2^n$)-gons inscribed in and circumscribed around a unit circle.  So $p(2)$ is the perimeter of a square inscribed in the unit circle.  $P(2)$ is the perimeter of a square circumscribed around a unit circle.

Archimedes proved the following (not in the formulaic language of algebra):  For all $n \geq 2$,

(P-formula)  $P(n+1) = \frac{2  p(n) P(n)}{p(n) + P(n)}$.

(p-formula)  $p(n+1) = \sqrt{ p(n)  P(n+1) }$.

1.  Compute $p(2)$ and $P(2)$.

2.  Use these formulas to compute $p(10)$ and $P(10)$.  Use this to get a good approximation for $\pi$!

We could use our previous sqrt function if you want, we'll take a fancier high-precision approach.  "mpmath" is a Python package for high-precision calculation.  It should be accessible through Google Colab.  You can read the full documentation at http://mpmath.org/doc/current/

First we load the package and print its status.

In [ ]:
from mpmath import *
print(mp)

Mpmath settings:
  mp.prec = 53                [default: 53]
  mp.dps = 15                 [default: 15]
  mp.trap_complex = False     [default: False]


The number mp.dps is (roughly) the number of decimal digits that mpmath will keep track of in its computations.  mp.prec is the binary precision, a bit more than 3 times the decimal precision.  We can change this to whatever we want.

In [ ]:
mp.dps = 50 # Let's try 50 digits precision to start.
print(mp)

Mpmath settings:
  mp.prec = 169               [default: 53]
  mp.dps = 50                 [default: 15]
  mp.trap_complex = False     [default: False]


mpmath has a nice function for square roots.  Compare this to your approximation from before!

In [ ]:
sqrt(2)  # mpf(...) stands for an mp-float.  

mpf('1.4142135623730951')

In [ ]:
type(sqrt(2)) # mpmath stores numbers in its own types!

mpmath.ctx_mp_python.mpf

In [ ]:
4*mp.atan(1) # mpmath has the arctan built in.  This should be pretty close to pi!

mpf('3.1415926535897932384626433832795028841971693993751068')

In [ ]:
# Combine Archimedes Section into one 
from mpmath import *
print(mp)

mp.dps = 50 # Let's try 50 digits precision to start.
print(mp)

sqrt(2)  # mpf(...) stands for an mp-float.  

type(sqrt(2)) # mpmath stores numbers in its own types!

4*mp.atan(1) # mpmath has the arctan built in.  This should be pretty close to pi!

Mpmath settings:
  mp.prec = 169               [default: 53]
  mp.dps = 50                 [default: 15]
  mp.trap_complex = False     [default: False]
Mpmath settings:
  mp.prec = 169               [default: 53]
  mp.dps = 50                 [default: 15]
  mp.trap_complex = False     [default: False]


mpf('3.1415926535897932384626433832795028841971693993751068')

Now try Archimedes' approximation of pi.  Use the mpmath sqrt function along the way.  How many iterations do you need to get pi correct to 100 digits?  Compare this to the arctan-series (not the mpmath atan function) via Machin's formula.

#### Solutions

Before creating our Python script to estimate `pi`, we needed to first understand that the perimeter of the square inside of a unit circle `p(2)` with a radius of `1` where `n` represents the a 2n-gon shape, is `sqrt(2)*4` while the perimeter of the square outside of a unit circle `P(2)` is `8`. 

We could now use `P(2)` and `p(2)` as our starting base point to compute for any 2n-gon, where `n >= 2`. With the given formulas, we could create a for loop that first calculates P(n+1), lets P(n) = P(n+1) for the next iteration, then calculates p(n+1), lets p(n) = p(n+1) for the next iteration, and finally, the for loop starts again with the next iteration until the loop has completed the given range. 

We have to note that `range(10)` does not mean that we will calculate `P(10)` or `p(n)` because technically, we are starting at `n=2`, not `n=1`. Thus, to find `P(10)` and `p(n)`, we would need to set the range at `range(8)`, giving us: `P(10) = 6.2832050205136178935273793169853200235845469162052` and `p(10) = 6.2831754505543194012577085254038374787985617171497`.

These values offer a rough estimation of the cirumference of the unit circle, indicating that we could need to let `P(10) = 2*pi*r` and `p(10) = 2*pi*r` where `pi = (P(10) / (2*r)` and `pi = (p(10) / (2*r)`. Thus, our approximation of `pi` when `P(10)` is `3.1416025102568089467636896584926600117922734581026
` while our approximation of `pi` when `p(10)` is `3.1415877252771597006288542627019187393992808585748`. 

In [44]:
#  Explore and experiment.
import math                                                       # Import math library for sqrt()
from mpmath import *                                              # Import mpmath library for digit precision

mp.dps = 50                                                       # Let's try 50 digits precision to start.

p_n = (4)*(sqrt(2))                                               # begin at a base of p(2) = 4*sqrt(2)
P_n = (2)*(4)                                                     # begin at a base of P(2) = 8

for i in range(8):                                                # takes 8 iterations to get p(10) and P(10)
  P_n_1 =  (2 * p_n * P_n) / (p_n + P_n)                          # P(n+1) = (2*p(n)*P(n))/(p(n) + P(n)) where p(n) and P(n) represents the perimeter of 2n-gon in the previous iteration; for P(3), we used P(2) and p(2)
  P_n = P_n_1                                                     # saves P(n+1) as P(n) for next iteration
  p_n_1 = sqrt(p_n * P_n_1)                                       # p(n+1) = sqrt(p(n)*P(n+1)) where p(n) was the previous iteration and P(n+1) is from current iteration; for p(3), we used p(2) and P(3)
  p_n = p_n_1                                                     # saves p(n+1) as p(n) for next iteration

print('After 8 iterations, we find that p(10) is: ' + str(p_n_1))                                                      
print('After 8 iterations, we find that p(10) is: ' + str(P_n_1))                                                      

print('Using the circumference of a circle formula and p(10), pi is approximated at: ' + str(p_n_1 / 2))                                                  
print('Using the circumference of a circle formula and P(10), pi is approximated at: ' + str(P_n_1 / 2))                                                  

After 8 iterations, we find that p(10) is: 6.2831754505543194012577085254038374787985617171497
After 8 iterations, we find that p(10) is: 6.2832050205136178935273793169853200235845469162052
Using the circumference of a circle formula and p(10), pi is approximated at: 3.1415877252771597006288542627019187393992808585748
Using the circumference of a circle formula and P(10), pi is approximated at: 3.1416025102568089467636896584926600117922734581026





In order for `pi` to be correct at `50` digits, we set our `mp.dps` to `50`and would need to implement more than just `8` recurions since `P(n) / (2) = p(n) / (2)` at `50` digits in order for us to conclude that `pi` is correct at `50` digits. Thereby, we completed a total of `84` iterations so that when `p(86) = P(86)`, `pi` is correct at `50` digits. 

In [36]:
#  Explore and experiment.
import math                                                       # Import math library for sqrt()
from mpmath import *                                              # Import mpmath library for digit precision

mp.dps = 50                                                       # Let's try 50 digits precision to start.

p_n = (4)*(sqrt(2))                                               # begin at a base of p(2) = 4*sqrt(2)
P_n = (2)*(4)                                                     # begin at a base of P(2) = 8

for i in range(84):                                               # takes 84 iterations to get "pi" at 50 digits precision 
  P_n_1 =  (2 * p_n * P_n) / (p_n + P_n)                          # P(n+1) = (2*p(n)*P(n))/(p(n) + P(n)) where p(n) and P(n) represents the perimeter of 2n-gon in the previous iteration; for P(3), we used P(2) and p(2)
  P_n = P_n_1                                                     # saves P(n+1) as P(n) for next iteration
  p_n_1 = sqrt(p_n * P_n_1)                                       # p(n+1) = sqrt(p(n)*P(n+1)) where p(n) was the previous iteration and P(n+1) is from current iteration; for p(3), we used p(2) and P(3)
  p_n = p_n_1                                                     # saves p(n+1) as p(n) for next iteration

print('After 84 iterations, we find that p(86) is: ' + str(p_n_1))                                                      
print('After 84 iterations, we find that p(86) is: ' + str(P_n_1))                                                      

print('Using the circumference of a circle formula and p(86), pi is approximated at: ' + str(p_n_1 / 2))                                                  
print('Using the circumference of a circle formula and P(86), pi is approximated at: ' + str(P_n_1 / 2))    

print('Because the difference between P(86) and p(86) stabilizes at: ' + str(abs((p_n_1 / 2) - (P_n_1 / 2))) + ', then we see that pi is correct at 50 digits.' )                            

After 84 iterations, we find that p(86) is: 6.2831853071795864769252867665590057683943387987502
After 84 iterations, we find that p(86) is: 6.2831853071795864769252867665590057683943387987502
Using the circumference of a circle formula and p(86), pi is approximated at: 3.1415926535897932384626433832795028841971693993751
Using the circumference of a circle formula and P(86), pi is approximated at: 3.1415926535897932384626433832795028841971693993751
Because the difference between P(86) and p(86) stabilizes at: 5.3455294201843912922810729343029637576303937602101e-51, then we see that pi is correct at 50 digits.



Note that `pi` is approximated to about 50 digits when we set our `mp.dps` to `50`. But suppose we want `pi` to be correct at `100` digits. Then we would need to change `mp.dps` to `100`and would need to implement more than just `89` recurions since `P(n) / (2) = p(n) / (2)` at `100` digits in order for us to conclude that `pi` is correct at `100` digits. Thereby, we completed a total of `166` iterations so that when `p(168) = P(168)`, `pi` is correct at `100` digits. 

In [37]:
#  Explore and experiment.
import math                                                       # Import math library for sqrt()
from mpmath import *                                              # Import mpmath library for digit precision

mp.dps = 100                                                       # Let's try 100 digits precision to start.

p_n = (4)*(sqrt(2))                                               # begin at a base of p(2) = 4*sqrt(2)
P_n = (2)*(4)                                                     # begin at a base of P(2) = 8

for i in range(168):                                              # takes 168 iterations to get "pi" at 100 digits precision 
  P_n_1 =  (2 * p_n * P_n) / (p_n + P_n)                          # P(n+1) = (2*p(n)*P(n))/(p(n) + P(n)) where p(n) and P(n) represents the perimeter of 2n-gon in the previous iteration; for P(3), we used P(2) and p(2)
  P_n = P_n_1                                                     # saves P(n+1) as P(n) for next iteration
  p_n_1 = sqrt(p_n * P_n_1)                                       # p(n+1) = sqrt(p(n)*P(n+1)) where p(n) was the previous iteration and P(n+1) is from current iteration; for p(3), we used p(2) and P(3)
  p_n = p_n_1                                                     # saves p(n+1) as p(n) for next iteration

print('After 168 iterations, we find that p(170) is: ' + str(p_n_1))                                                      
print('After 168 iterations, we find that p(170) is: ' + str(P_n_1))                                                      

print('Using the circumference of a circle formula and p(170), pi is approximated at: ' + str(p_n_1 / 2))                                                  
print('Using the circumference of a circle formula and P(170), pi is approximated at: ' + str(P_n_1 / 2))    

print('Because the difference between P(170) and p(170) stabilizes at: ' + str(abs((p_n_1 / 2) - (P_n_1 / 2))) + ', then we see that pi is correct at 100 digits.' )                            

After 168 iterations, we find that p(170) is: 6.283185307179586476925286766559005768394338798750211641949889184615632812572417997256069650684234136
After 168 iterations, we find that p(170) is: 6.283185307179586476925286766559005768394338798750211641949889184615632812572417997256069650684234136
Using the circumference of a circle formula and p(170), pi is approximated at: 3.141592653589793238462643383279502884197169399375105820974944592307816406286208998628034825342117068
Using the circumference of a circle formula and P(170), pi is approximated at: 3.141592653589793238462643383279502884197169399375105820974944592307816406286208998628034825342117068
Because the difference between P(170) and p(170) stabilizes at: 0.0, then we see that pi is correct at 100 digits.



Compared to the `arctan` method, while the `arctan` method was quicker, we are a bit more limited to what we could do because those values are more exact and we cannot alter those `arctan` values as much as we like. 

With Archimedes' Approximation method, we can increase our precision settings and the number of iterations so that we could gain a more accurate and precise value of `pi`, thus this method is more versatile than the `arctan` method when we want more accuracy and more digit approximations. 